crates/compiler/mono/src/inc_dec.rs

```diff
@@ -1,1554 +1,1193 @@
                     }
-                }
-
-                stmt
-            }};
-        }
-
-        // incrementing/consuming the closure (if needed) is done by the zig implementation. We
-        // don't want to touch the RC on the roc side, so treat these as borrowed.
-        const FUNCTION: bool = BORROWED;
-        const CLOSURE_DATA: bool = BORROWED;
-
-        let borrows = [FUNCTION, CLOSURE_DATA];
-        let after_arguments = &arguments[op.function_index()..];
-
-        match *op {
-            ListMap { xs } => {
-                let ownerships = [(xs, function_ps[0])];
-
-                let b = self.add_dec_after_lowlevel(after_arguments, &borrows, b, b_live_vars);
-
-                let b = handle_ownerships_post!(b, ownerships);
-
-                let v = create_call!(function_ps.get(1));
-
-                handle_ownerships_pre!(Stmt::Let(z, v, l, b), ownerships)
-            }
-            ListMap2 { xs, ys } => {
-                let ownerships = [(xs, function_ps[0]), (ys, function_ps[1])];
-
-                let b = self.add_dec_after_lowlevel(after_arguments, &borrows, b, b_live_vars);
-
-                let b = handle_ownerships_post!(b, ownerships);
-
-                let v = create_call!(function_ps.get(2));
 
-                handle_ownerships_pre!(Stmt::Let(z, v, l, b), ownerships)
-            }
-            ListMap3 { xs, ys, zs } => {
-                let ownerships = [
-                    (xs, function_ps[0]),
-                    (ys, function_ps[1]),
-                    (zs, function_ps[2]),
-                ];
-
-                let b = self.add_dec_after_lowlevel(after_arguments, &borrows, b, b_live_vars);
-
-                let b = handle_ownerships_post!(b, ownerships);
-
-                let v = create_call!(function_ps.get(3));
+                    consume_symbols
+                }
+            };
 
-                handle_ownerships_pre!(Stmt::Let(z, v, l, b), ownerships)
-            }
-            ListMap4 { xs, ys, zs, ws } => {
-                let ownerships = [
-                    (xs, function_ps[0]),
-                    (ys, function_ps[1]),
-                    (zs, function_ps[2]),
-                    (ws, function_ps[3]),
-                ];
+            // Given the consume_symbols we can determine what additional symbols should be dropped in each branch.
+            let newer_branches = Vec::from_iter_in(
+                new_branches
+                    .into_iter()
+                    .map(|(label, info, branch, branch_env)| {
+                        // If the symbol is owned in the branch, it is not used in the branch and we can drop it.
+                        let consume_symbols_branch = consume_symbols.iter().copied().filter(|consume_symbol| {
+                             matches!(branch_env.get_symbol_ownership(consume_symbol).expect("All symbols defined in the current environment should be in the environment of the branches."), Ownership::Owned) 
```

Comment:
the long error strings break formatting here. a separate `let msg = "All ..." `  would at least help visually